prototypes/ScheduledMergesTest.hs

```diff
@@ -335,34 +336,56 @@ instance Arbitrary T where
       , length ts' > 1
       ]
 
-instance Arbitrary I where
-  arbitrary = QC.oneof [ISingle <$> arbitrary, IMerging <$> arbitrary]
-  shrink (ISingle r)  = [ISingle r' | r' <- shrink r]
-  shrink (IMerging m) = [ISingle (completeM m)] <> [IMerging m' | m' <- shrink m]
+instance Arbitrary P where
+  arbitrary = QC.oneof [PRun <$> arbitrary, PMergingRun <$> arbitrary]
+  shrink (PRun r)        = [PRun r' | r' <- shrink r]
+  shrink (PMergingRun m) = [PRun (completeM m)]
+                        <> [PMergingRun m' | m' <- shrink m]
 
 instance (Arbitrary t, IsMergeType t) => Arbitrary (M t) where
   arbitrary = QC.frequency
-      [ (1, MCompleted <$> arbitrary <*> arbitrary)
+      [ (1, do (mt, r) <- arbitrary
+               let md = MergeDebt (runSize r)
+               pure $ MCompleted mt md r)
       , (1, do
           mt <- arbitrary
-          n <- QC.chooseInt (2, 8)
+          n  <- QC.chooseInt (2, 8)
           rs <- QC.vectorOf n (QC.scale (`div` n) arbitrary)
-          let totalWork = sum (map (length . getNonEmptyRun) rs)
-          workRemaining <- QC.chooseInt (1, totalWork)
-          unspentCredits <- QC.chooseInt (0, min mergeBatchSize workRemaining - 1)
-          let d = MergeDebt unspentCredits workRemaining
-          return (MOngoing mt d rs))
+          let totalDebt    = sum (map (length . getNonEmptyRun) rs)
+          suppliedCredits <- QC.chooseInt (0, totalDebt-1)
+          unspentCredits  <- QC.chooseInt (0, min (mergeBatchSize-1) suppliedCredits)
+          let spentCredits = suppliedCredits - unspentCredits
+          let md = MergeDebt {
+                     totalDebt
+                   }
+              mc = MergeCredit {
+                    unspentCredits,
+                    spentCredits
+                 }
+          assert (mergeDebtInvariant md mc) $
+            return (MOngoing mt md mc rs))
```

Comment:
I think at this point it would be worth splitting out functions `genMergeCreditForDebt` and `shrinkMergeCreditForDebt` that take care of all the details and invariants.